### PR TITLE
[FrameworkBundle] Add new "controller.service_arguments" tag to inject services into actions

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | (no test yet)
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

Talking with @simensen and @weaverryan, we wondered if we could leverage the `ArgumentResolver` mechanism to make it inject services on demand, using e.g. autowiring.

```php
class PostController
{
  public function indexAction(Request $request, PostRepository $postRepository)
  {
    // PostRepository comes from the container
    $postRepository->findAll(); // ...
  }
}
```

This PR achieves that, using a new "controller.service_arguments" tag. Typically:
```yaml
services:
    AppBundle\Controller\PostController:
        autowire: true
        tags:
            - name: controller.service_arguments
```

It also supports with explicit wiring (thus doesn't necessarily require autowiring if you don't want to use it):
```yaml
services:
    AppBundle\Controller\PostController:
        tags:
            - name: controller.service_arguments
              action: fooAction
              argument: logger
              id: my_logger
```

~~The attached diff is bigger than strictly required for now, until #21770 is merged.~~

Todo:
- [x] rebase on top of #21770 when merged
- [x] add tests
- [x] add cleaning pass to remove empty service locators

### CODE REVIEW

**Comment by iltar** [comment]
I'm personally against this suggestion, especially now that autowiring the constructor is a lot easier: https://github.com/symfony/symfony/pull/21164#issuecomment-270857003

> If you ask me:
> - **Constructor** stateless dependencies, e.g. services, things that are the same no matter what action is called.
> - **Action** stateful dependencies, e.g. entities, parameters or other objects that can have different states each time it gets called.
>
>The session can be different each time it's called, therefore it shouldn't be a constructor dependency. However, when using the `Request::getSession()`, you don't have access to `getFlashBag()`, which is `Session` specific. 

I think that adding the ability to "inject" services into the action arguments, will add a layer of magic that makes it more complex in combination with autowiring.

**Comment by nicolas-grekas** [comment]
Now with explicit wiring:
`<tag name="routing.controller" action="fooAction" argument="logger" service="my_logger"/>`
repeat the tag for several args/actions.

And with tests.
PR is thus ready.
Status: needs review

@iltar: except when there is only one action (e.g. `__invoke` style - but this is not the general case), constructor injection is not an option for controllers, because this kind of injection misses a core property that is required here: laziness.

That's why people inject the container instead in fact (and/or extend the base controller really).
We all know that injecting the container is an anti-pattern - the bad kind of service locators.

By injecting services into actions, we get laziness, and we ask people for writing a type hints instead of calling `$this->get('foo')`. This makes dependencies explicit - thus helps the reader, and the IDE (auto-completion without the clever phpstorm plugin for Symfony). DX wise, it's a good trade - an enjoyable one to me.

The alternatives we have that provide laziness are getter injection - and [service subscribers](https://github.com/symfony/symfony/pull/21708). But these won't provide the same DX here (getter injection is nice when shipped via a trait, see #18193).

All in all, this will provide a seamless DX to me, without making people write "bad" code anymore.

**Comment by iltar** [comment]
Will this make it possible to do something like `<tag name="routing.controller">` and have it registered as service? This would replace the annotation for me:
```php
/**
 * @Route(service="...")
 */
class FooController {}
```

**Comment by nicolas-grekas** [comment]
@iltar so, this is good to you now? cool :)
About your question, this won't help (#21282 could). Maybe `ContainerControllerResolver` could do a by-fqcn lookup in the container? Dunno if that'd be a good idea thought. I'll let you give it a try if you want :)

**Comment by iltar** [comment]
@nicolas-grekas I'm not against it and your argument makes sense!

**Comment by simensen** [comment]
I need to look at the actual implementation in more detail, but what is described here sounds pretty close to what I was hoping for in this implementation.

I also like this bit:

```
<tag name="routing.controller" action="fooAction" argument="logger" service="my_logger"/>
```

What would I want to do if I wanted to have all actions mapped automatically and also map a specific service? Would it work like this?

```
<tag name="routing.controller" action="*Action"/>
<tag name="routing.controller" action="fooAction" argument="logger" service="my_logger"/>
```

... or even:

```
<tag name="routing.controller" action="fooAction"/>
<tag name="routing.controller" action="fooAction" argument="logger" service="my_logger"/>
```

How would `__invoke` be handled? Would it be possible to assume `__invoke` if `action` is not specified? For example, if you had a simple controller that only had `__invoke`, you could do this?

```
<tag name="routing.controller"/>
```

All-in-all, I think I like this approach. I know you had some technical reasons why it might not be the best way to go. Were there any things you've done in this PR that you wish you could have done differently?

Tentative :+1: from me for now. :) Super excited about this direction!

**Comment by nicolas-grekas** [comment]
@simensen:

> Would it work like this?
```
<tag name="routing.controller" action="*Action"/>
<tag name="routing.controller" action="fooAction" argument="logger" service="my_logger"/>
```
Yes! You'd just need one more thing to have the DX you envision: enable autowiring on the definitions that have these tags. The "routing.controller" does not enable autowiring automatically (that'd be coupling), but it plays well with autowired definitions (that's composition).

```
<tag name="routing.controller" action="fooAction"/>
<tag name="routing.controller" action="fooAction" argument="logger" service="my_logger"/>
```
That one would be redundant, only the second tag is needed.

> Would it be possible to assume `__invoke` if `action` is not specified?

As you know, I wondered the same. Thinking about this, it doesn't make much sense for `__invoke`-style controllers to use the "routing.controller" tag: they could very well use classical constructor injection. That'd be what I'd recommend in fact.
As such, the "action" attribute is required. An exception is thrown if it's missing.

> you had some technical reasons why it might not be the best way to go.

They vanished now that we have #21770, which serves as the infrastructure that this PR leverages.

I'm really happy with the PR as is :)

**Comment by simensen** [comment]
@nicolas-grekas:

> I'm really happy with the PR as is :)

That makes *me* happy. :)

I'm a bit lost on this:

> enable autowiring on the definitions that have these tags

Does this mean that if autowiring is defined for the controller in question then autowiring is used but if not then it isn't?

> ... it doesn't make much sense for __invoke-style controllers to use the "routing.controller" tag: they could very well use classical constructor injection. That'd be what I'd recommend in fact.

What about closures? Or is that something we don't need to worry about supporting in core? For example, Silex-style closures:

```
$app->get('/', function (Request $request, PostRepository $postRepository) {
    // ...
});
```

This might be well out of scope so I'm not going to worry about this too much. Mostly just curious if this would somehow be covered?

**Comment by nicolas-grekas** [comment]
> if autowiring is defined for the controller in question then autowiring is used but if not then it isn't?

yep

> What about closures?

All of this is bound to Symfony's DIC (almost all the logic is in a compiler pass), so Silex would need its own wiring system.

**Comment by simensen** [comment]
> All of this is bound to Symfony's DIC (almost all the logic is in a compiler pass), so Silex would need its own wiring system.

Got it. Makes sense.

---

👍

**Comment by nicolas-grekas** [comment]
I just found two edges to polish:
- magic php methods (all those starting with double underscore) are now skipped by wildcard actions
- the service value resolver is unregistered if there is no controllers using the tag at all

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/ServiceValueResolver.php [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+    {
+        $this->container = $container;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function supports(Request $request, ArgumentMetadata $argument)
+    {
+        return is_string($controller = $request->attributes->get('_controller')) && $this->container->has($controller) && $this->container->get($controller)->has($argument->getName());
```
Comment (HeahDude) [inline src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/ServiceValueResolver.php]: Is there any way to optimize the calls? AFAIU `$request->attributes->get('_controller')`, `$this->container->has($controller)`, `$this->container->get($controller)->has($argument->getName()` will be called for every arguments of the controller although we could keep a map of controllers by request that can be reused even in `resolve()`, wdyt?

Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/ServiceValueResolver.php]: The calls resolve to immediate isset checks, can't be faster

**Comment by nicolas-grekas** [comment]
Question to reviewers: should the pass be moved to HTTP kernel? The tag be renamed ?

**Comment by weaverryan** [comment]
Hi guys!

This is really interesting! I think the approach of putting the arguments for each controller into its own ServiceLocator service is super clever and powerful!

I think this PR may contain the best possible way to implement this feature. But, it still may not be good enough, because it's error prone. I mean, it's too easy for the user to make an error. And that error will be have a clear message.

For example,, if you do everything correct, but configure your route to *not* use the service syntax (e.g. you use `AppBundle:Foo:bar`, or you use `@Route`, without explicitly specifying the service id above the controller). Or, you do everything right, use `action: *Action`, but then accidentally forget to suffix your method with `Action` (e.g. because you're using `@Route`... where this is easy to do and perfectly legal). In these cases, instead of the user getting a clear, compile-time error, they'll get a strange run-time error - i.e. `controller requires that you provide a value for the $logger argument...` - but no indication as to *why* this is not working. It might even be working for one controller, but not for another.

Backing up, in a perfect world, this argument resolver would work for *all* controllers, regardless of how they are registered or used. This is how all the other argument resolvers work. The big problem is that the type-hint => service resolution (i.e. autowiring) is only done during container compile time. But controllers are only ever known at runtime. Sure, we could even read the routing configuration at compile time to try to get a list of all of the controllers, but even that falls short: it wouldn't catch controllers used in sub-requests (e.g. `render(controller())` in Twig or forwarding).

tl;dr The implementation of *how* we determine the service arguments for a controller method, store them and look them up is not the problem - it's pretty brilliant actually. The problem is that there is no decent way to know all of the controllers in the system at compile time. And unless we can figure that out, I don't think this feature works. 

Thoughts & ideas?

**Comment by iltar** [comment]
@weaverryan all metadata required is known, so regarding error messages, it would need some extra inspection when throwing the exception to determine a more developer friendly error.

https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpKernel/Controller/ArgumentResolver.php#L50-L79

File: src/Symfony/Component/Routing/DependencyInjection/RoutingControllerPass.php [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+        $serviceResolver = $container->getDefinition($this->resolverServiceId);
+        $parameterBag = $container->getParameterBag();
+        $controllers = array();
+
+        foreach ($container->findTaggedServiceIds($this->controllerTag) as $id => $tags) {
+            $def = $container->getDefinition($id);
+            $class = $def->getClass();
+            $isAutowired = $def->isAutowired();
+
+            if ($def->isAbstract()) {
```
Comment (iltar) [inline src/Symfony/Component/Routing/DependencyInjection/RoutingControllerPass.php]: assigning `$class` and `$isAutowired should probably be done below this if check, as they are not needed before it.

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/ServiceValueResolver.php (hunk 1 above)
Comment (iltar) [inline src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/ServiceValueResolver.php]: What if I would use the other notation here? `Foo::Bar`, with controller as a service, this would be supported.

Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/ServiceValueResolver.php]: @iltar it will work: both variants are now registered as references when id==class

Comment (iltar) [inline src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/ServiceValueResolver.php]: Ah yes, you're right! https://github.com/symfony/symfony/pull/21771/files#diff-4f6a92ae3706365662df9167009eaffaR205

**Comment by nicolas-grekas** [comment]
PR updated. Everything now moved to `HttpKernel`, tag renamed to `controller.service_arguments`.

When the tag is set, all public methods are now considered for generating a service-locator based argument-resolver.
This means the "action" attribute is now optional. If provided, it doesn't accept wildcards anymore. It must now be accompanied by the "argument" and "service" attributes. Eg same as before:
`<tag name="controller.service_arguments" action="fooAction" argument="logger" service="my_logger"/>`

All public methods are considered for generating a service-locator: in fact, methods that are listed to be called at instantiation time (ie the one returned by `Definition::getMethodCalls()`) are excluded of course.

So, we now have two passes:
- the first looks like a restricted version of the previous one (no wildcard support, all methods considered as service-locator generator candidates) - it generates the service locators;
- the second one runs just before removing passes, and removes any emptied service-locators (because of "ignore-on-invalid" processing), and any service-locators bound to call-at-instantiation methods.

There is a possibility that this may generate some unused service-locators, bound to public methods that are not used as actions. Yet, that should be rare and that'll be just dead code with no practical downside, especially compared to the huge +-side of not having to configure anything.

> if you do everything correct, but configure your route to not use the service syntax

this PR provides nothing new on this topic: today also, if you add a bogus argument to an action (eg a bad type hint for a doctrine entity with the proper argument resolver configured), then the controller fails when it called without earlier notice. There is no chance to make anything better here, at least that's not what this PR is about - entities or services or whatever.

> use action: *Action, but then accidentally forget to suffix your method with Action

valid concern, fixed! see description above.

> in a perfect world, this argument resolver would work for all controllers

not in *my* perfect world! I certainly do not want to have services come in when I did not ask for!
That'd mean we'd rely on some global mechanism that knows better then me *which* services should be injected. One may think that, well, just do `$container->get(MyTypeHint::class)`, or `$autowirer->getTheServiceThatMatches(LoggerInterface::class)`. If so, please realize that this is applying a convention that you can't escape from. Symfony has always provided the opposite, via flexible OOP architectures that you can use however you want/need (no hardcoded convention).

> The problem is that there is no decent way to know all of the controllers in the system at compile time.

As stated, that's certainly not a problem to me, quite the contrary.
Keeping things simple means *not* doing that kind of global things.
Instead, Symfony has always chosen the "be explicit" path.
No exception here.

> when the user makes a mistake, we must tell them exactly what went wrong - no wtf moments

I really like that approach, thanks for pushing it to us.
So,  this topic, the code has may checks/exceptions thrown in each possible bad situations.
And in situations where things cannot be decided as legitimate or mistake, I added log messages so that the user can inspect what's going on by looking at them.

@weaverryan, I think I addressed all your concerns :)

**Comment by mvrhov** [comment]
Instead of that. Why not promote one Action per class. Then everything works with constructor autowiring which is already in core

**Comment by nicolas-grekas** [comment]
"service" attribute renamed to "id" for consistency with existing practice.
@mvrhov classical actions are still first class citizens.

**Comment by weaverryan** [comment]
Awesome! So this new approach is very interesting - it removes several of the issues I was having earlier!

Before we dive further into the details, the implementation makes me realize something: from a *purely* technical perspective, this feature could work *without* the controllers being services. If I told you that I wanted service auto-completion on my `AppBundle\Controller\SecurityController` actions, then you could create a service locator for all of those arguments. After all, the `RegisterControllerArgumentLocatorsPass` does not actually rely on the controller being a service... other than this is the way you opt into this feature. But (again, from a purely technical perspective... just stay with me), it would be equally possible to allow the user to opt into this feature without needing to register the controller as a service. For example:

```yml
framework:
    controllers_service_arguments:
        AppBundle\Controller: true
        # .. and you could probably have config to control specific args...
```

I'm not proposing this is a superior syntax! It just strikes me as a bit unnatural that we're requiring the controllers to be services... when we don't really need that! In fact, one of my original comments was this:

> if you do everything correct, but configure your route to not use the service syntax, then the feature doesn't work.

You said you didn't do anything to address this... but you did! Thanks to this line in `RegisterControllerArgumentLocatorsPass`:

```php
if ($id === $class) {
    $controllers[$id.'::'.$r->name] = new Reference($argsId);
}
```

It is now possible to register a controller as a service via the PSR-4 loader, but then refer to it using the non-service syntax. Yes: as long as your register the controller as a service with the correct tag, you are then allowed to use it NOT as a service, but enjoy the service argument injection. I can imagine this conversation:

> Person 1) I can't get Symfony to auto-inject my LoggerInterface argument to my controller
> Person 2) Just register the controllers as services, and it'll work

But in reality, Person 1 is still not actually using their controller as a service. This whole situation is what made me realize that the requirement that the controllers be services is unnatural... simply because the feature itself doesn't require it! Even if everyone was creating all controllers as services, I still think the implementation shouldn't rely on this unrelated thing.

So:

1) I 100% agree that you should be able to opt into / opt out of this feature (and somehow control edge case services - i.e. the `action`, `id`, `argument` options in your implementation)

2) But, since the implementation doesn't rely on controllers being services, I think the opt-in should be done in a different, but similar way - i.e. by pointing at what classes you want this for. Then, the controller names in the service locator would be keyed almost exactly like now: using the `className::methodName` syntax (`arguments.AppBundle\Controller\SecurityController::loginAction`) and also any service ids that match the class name (e.g. `arguments.AppBundle\Controller\SecurityController:loginAction` or `security_controller:loginAction` if I created a custom service id).

What are your thoughts about this? And thanks :)

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Comment by nicolas-grekas** [comment]
Happy the PR is merged, let's play with it now :)

@weaverryan to answer your comment, having left some time to think about it:

> this feature could work without the controllers being services. If I told you that I wanted service auto-completion on my AppBundle\Controller\SecurityController actions

The PR just needs a service locator that returns something for any string value of the  "_controller" routing param. So yes this can work with, any such "_controller", not only those registered as services.
Yet as you spotted, this service locator needs to be created.

I think it would be possible to implement this "controllers_service_arguments" config key yes. I just don't see the need to fight against having controllers-as-services, thus don't see the need to seek for any other configuration mean.

> Person 1) I can't get Symfony to auto-inject my LoggerInterface argument to my controller
> Person 2) Just register the controllers as services, and it'll work

I don't see anything wrong with P2's answer. We need a "Px" that will need to answer something to P1, because P1 needs to opt-in. Being via service registration or another mean doesn't matter.
In fact, since we know this will always be possible via services, seeking for a P3 will just create alternatives, thus make things confusing ("should I follow P2 or P3?!")

I really look forward the RX (reviewer experience) where I don't need do find `$this->get('foo')` nested in the code anymore, but just have to look at the action signature to know what and how services are going to be used.

**Comment by fabpot** [comment]
Just tested it on a project... and that works! It's deceptively simple and I will activate it in the default configuration for apps using Flex. Can't wait to share that with everyone.

**Comment by weaverryan** [comment]
Docs issue added! https://github.com/symfony/symfony-docs/issues/7672

File: src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+    public function __construct($resolverServiceId = 'argument_resolver.service', $controllerTag = 'controller.service_arguments')
+    {
+        $this->resolverServiceId = $resolverServiceId;
+        $this->controllerTag = $controllerTag;
+    }
+
+    public function process(ContainerBuilder $container)
+    {
+        if (false === $container->hasDefinition($this->resolverServiceId)) {
```
Comment (HeahDude) [inline src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php]: Why not the other way around? If some tags are found then register the resolver.

Because currently the resolver is asked to resolve before DefaultValueResolver even if there is no controller tagged.

Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php]: I'd say because "less dynamic" is better. The extra check you're talking about is not a perf issue: it is really fast - and happens onces per action (you don't call that many actions for constructing one response, do you?)

Comment (HeahDude) [inline src/Symfony/Component/HttpKernel/DependencyInjection/RegisterControllerArgumentLocatorsPass.php]: >and happens once per action

Isn't it once per argument of an action, and sub-action(s)? Although I'm fine with it.

**Comment by GeoffreyHervet** [comment]
In symfony4, will you remove `\Symfony\Bundle\FrameworkBundle\Controller\Controller` ?
